README.md
# Flutter VoIP Push Notification

[![pub package](https://img.shields.io/pub/v/flutter_voip_push_notification.svg)](https://pub.dartlang.org/packages/flutter_voip_push_notification)
Flutter VoIP Push Notification - Currently iOS >= 8.0 only

## Motivation

Since iOS 8.0 there is an execellent feature called **VoIP Push Notification** ([PushKit][1]), while [firebase_messaging][3] does not support voip push notification which is only available on iOS >= 8.0 which is the reason for this plugin.

To understand the benefits of **Voip Push Notification**, please see [VoIP Best Practices][2].

**Note 1**: This plugin works for only iOS. You can use [firebase_messaging][3] for Android by [sending high priority push notification][5]

**Note 2** This This plugin was inspired by [react-native-voip-push-notification][4] and [firebase_messaging][3]

### iOS

The iOS version should be >= 8.0 since we are using [PushKit][1].

#### Enable VoIP Push Notification and Get VoIP Certificate

Please refer to [VoIP Best Practices][2].

**Note**: Do NOT follow the `Configure VoIP Push Notification` part from the above link, use the instruction below instead.

#### AppDelegate.swift

```swift

...

import PushKit                     /* <------ add this line */
import flutter_voip_push_notification      /* <------ add this line */
...

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate, PKPushRegistryDelegate {

    ...

    /* Add PushKit delegate method */

    // Handle updated push credentials
    func pushRegistry(_ registry: PKPushRegistry,
                      didReceiveIncomingPushWith payload: PKPushPayload,
                      for type: PKPushType,
                      completion: @escaping () -> Void){
        // Register VoIP push token (a property of PKPushCredentials) with server
        FlutterVoipPushNotificationPlugin.didReceiveIncomingPush(with: payload, forType: type.rawValue)
    }

    // Handle invalidated push tokens
    func pushRegistry(_ registry: PKPushRegistry, didInvalidatePushTokenFor type: PKPushType) {
        FlutterVoipPushNotificationPlugin.didInvalidatePushToken(forType: type.rawValue)
    }

    // Handle incoming pushes
    func pushRegistry(_ registry: PKPushRegistry, didUpdate pushCredentials: PKPushCredentials, for type: PKPushType) {
        // Process the received push
        FlutterVoipPushNotificationPlugin.didUpdate(pushCredentials, forType: type.rawValue);
    }

    ...
}
```

#### AppDelegate.m Modification

```objective-c

...

#import <PushKit/PushKit.h>                    /* <------ add this line */
#import "FlutterVoipPushNotificationPlugin.h"      /* <------ add this line */

...

@implementation AppDelegate

- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{

...

/* Add PushKit delegate method */

// Handle updated push credentials
- (void)pushRegistry:(PKPushRegistry *)registry didUpdatePushCredentials:(PKPushCredentials *)credentials forType:(NSString *)type {
  // Register VoIP push token (a property of PKPushCredentials) with server
  [FlutterVoipPushNotificationPlugin didUpdatePushCredentials:credentials forType:(NSString *)type];
}

// The system calls this method when a previously provided push token is no longer valid for use. No action is necessary on your part to reregister the push type. Instead, use this method to notify your server not to send push notifications using the matching push token.
- (void)pushRegistry:(PKPushRegistry *)registry didInvalidatePushTokenForType:(PKPushType)type
{
  [FlutterVoipPushNotificationPlugin didInvalidatePushTokenForType:(NSString *)type];
}

// Handle incoming pushes
- (void)pushRegistry:(PKPushRegistry *)registry didReceiveIncomingPushWithPayload:(PKPushPayload *)payload forType:(NSString *)type {
  // Process the received push
  [FlutterVoipPushNotificationPlugin didReceiveIncomingPushWithPayload:payload forType:(NSString *)type];
}

...

@end

```

## Usage

Add `flutter_voip_push_notification` as a [dependency in your pubspec.yaml file](https://flutter.io/using-packages/).

### Example

```dart

import 'package:flutter/material.dart';
import 'dart:async';
import 'package:flutter_voip_push_notification/flutter_voip_push_notification.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _pushToken = '';
  FlutterVoipPushNotification _voipPush = FlutterVoipPushNotification();
  @override
  void initState() {
    super.initState();
    configure();
  }

  // Configures a voip push notification
  Future<void> configure() async {
    // request permission (required)
    await _voipPush.requestNotificationPermissions();

    // listen to voip device token changes
    _voipPush.onTokenRefresh.listen(onToken);

    // do configure voip push
    _voipPush.configure(
      onMessage: onMessage,
      onResume: onResume,
      onInvalidToken: onInvalidToken,
    );
  }

  /// Called when the device token changes
  void onToken(String token) {
    // send token to your apn provider server
    setState(() {
      _pushToken = token;
    });
  }

  /// Called to receive notification when app is in foreground
  ///
  /// [isLocal] is true if its a local notification or false otherwise (remote notification)
  /// [payload] the notification payload to be processed. use this to present a local notification
  Future<dynamic> onMessage(bool isLocal, Map<String, dynamic> payload) {
    // handle foreground notification
    print("received on foreground payload: $payload, isLocal=$isLocal");
    return null;
  }

  /// Called to receive notification when app is resuming from background
  ///
  /// [isLocal] is true if its a local notification or false otherwise (remote notification)
  /// [payload] the notification payload to be processed. use this to present a local notification
  Future<dynamic> onResume(bool isLocal, Map<String, dynamic> payload) {
    // handle background notification
    print("received on background payload: $payload, isLocal=$isLocal");
    showLocalNotification(payload);
    return null;
  }

  /// Call to receive an PushKit invalid token
  ///
  /// [invalidToken] is the token that is no longer valid
  /// Check out why a token could no longer be valid
  /// https://stackoverflow.com/questions/46977380/voip-push-under-what-circumstances-does-didinvalidatepushtokenfortype-get-calle#47015401
  Future<dynamic> onInvalidToken(String invalidToken) {
    // Tell the server to remove the invalid token
    print("received on background invalidToken: $invalidToken");
    return null;
  }

  Future<void> showLocalNotification(Map<String, dynamic> notification) {
    String alert = notification["aps"]["alert"];
    return _voipPush.presentLocalNotification(LocalNotification(
      alertBody: "Hello $alert",
    ));
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text('Received Voip Push token: $_pushToken\n'),
        ),
      ),
    );
  }
}

```

[1]: https://developer.apple.com/library/ios/documentation/NetworkingInternet/Reference/PushKit_Framework/index.html
[2]: https://developer.apple.com/library/ios/documentation/Performance/Conceptual/EnergyGuide-iOS/OptimizeVoIP.html
[3]: https://github.com/FirebaseExtended/flutterfire/tree/master/packages/firebase_messaging
[4]: https://github.com/react-native-webrtc/react-native-voip-push-notification
[5]: https://developers.google.com/cloud-messaging/concept-options#setting-the-priority-of-a-message


CHANGELOG.md
# Flutter Voip Push Notification

## 0.0.4

- Add support for invalidating tokens

## 0.0.3

- Rename IOSNotificationSettings to NotificationSettings and refactored example

## 0.0.2

- Add support for showing local notification, uses app delegate as push delegate

## 0.0.1

- Initial release.


example/test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility that Flutter provides. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:flutter_voip_push_notification_example/main.dart';

void main() {
  testWidgets('Verify Token Received', (WidgetTester tester) async {
    // Build our app and trigger a frame.
    await tester.pumpWidget(MyApp());

    // Verify that voip push token is received
    expect(
      find.byWidgetPredicate(
        (Widget widget) => widget is Text &&
                           widget.data.startsWith('Received Voip Push token received:'),
      ),
      findsOneWidget,
    );
  });
}


test/flutter_voip_push_notification_test.dart
import 'package:flutter/services.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:flutter_voip_push_notification/flutter_voip_push_notification.dart';

void main() {
  const MethodChannel channel = MethodChannel('flutter_voip_push_notification');

  TestWidgetsFlutterBinding.ensureInitialized();

  setUp(() {
    channel.setMockMethodCallHandler((MethodCall methodCall) async {
      return '42';
    });
  });

  tearDown(() {
    channel.setMockMethodCallHandler(null);
  });

  test('getPlatformVersion', () async {
    expect(await FlutterVoipPushNotification.platformVersion, '42');
  });
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'dart:async';
import 'package:flutter_voip_push_notification/flutter_voip_push_notification.dart';

void main() => runApp(MyApp());

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  String _pushToken = '';
  FlutterVoipPushNotification _voipPush = FlutterVoipPushNotification();
  @override
  void initState() {
    super.initState();
    configure();
  }

  // Configures a voip push notification
  Future<void> configure() async {
    // request permission (required)
    await _voipPush.requestNotificationPermissions();

    // listen to voip device token changes
    _voipPush.onTokenRefresh.listen(onToken);

    // do configure voip push
    _voipPush.configure(
      onMessage: onMessage,
      onResume: onResume,
      onInvalidToken: onInvalidToken,
    );
  }

  /// Called when the device token changes
  void onToken(String token) {
    // send token to your apn provider server
    setState(() {
      _pushToken = token;
    });
  }

  /// Called to receive notification when app is in foreground
  ///
  /// [isLocal] is true if its a local notification or false otherwise (remote notification)
  /// [payload] the notification payload to be processed. use this to present a local notification
  Future<dynamic> onMessage(bool isLocal, Map<String, dynamic> payload) {
    // handle foreground notification
    print("received on foreground payload: $payload, isLocal=$isLocal");
    return null;
  }

  /// Called to receive notification when app is resuming from background
  ///
  /// [isLocal] is true if its a local notification or false otherwise (remote notification)
  /// [payload] the notification payload to be processed. use this to present a local notification
  Future<dynamic> onResume(bool isLocal, Map<String, dynamic> payload) {
    // handle background notification
    print("received on background payload: $payload, isLocal=$isLocal");
    showLocalNotification(payload);
    return null;
  }

  /// Call to receive an PushKit invalid token
  ///
  /// [invalidToken] is the token that is no longer valid
  /// Check out why a token could no longer be valid
  /// https://stackoverflow.com/questions/46977380/voip-push-under-what-circumstances-does-didinvalidatepushtokenfortype-get-calle#47015401
  Future<dynamic> onInvalidToken(String invalidToken) {
    // Tell the server to remove the invalid token
    print("received on background invalidToken: $invalidToken");
    return null;
  }

  Future<void> showLocalNotification(Map<String, dynamic> notification) {
    String alert = notification["aps"]["alert"];
    return _voipPush.presentLocalNotification(LocalNotification(
      alertBody: "Hello $alert",
    ));
  }

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Plugin example app'),
        ),
        body: Center(
          child: Text('Received Voip Push token: $_pushToken\n'),
        ),
      ),
    );
  }
}


example/ios/Runner/AppDelegate.swift
import UIKit
import Flutter
import PushKit
import CallKit
import flutter_voip_push_notification

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate, PKPushRegistryDelegate {
    
    override func application(
        _ application: UIApplication,
        didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
        ) -> Bool {
        GeneratedPluginRegistrant.register(with: self)
        return super.application(application, didFinishLaunchingWithOptions: launchOptions)
    }
    
    /* Add PushKit delegate method */
    
    // Handle updated push credentials
    func pushRegistry(_ registry: PKPushRegistry,
                      didReceiveIncomingPushWith payload: PKPushPayload,
                      for type: PKPushType,
                      completion: @escaping () -> Void){
        
        // Register VoIP push token (a property of PKPushCredentials) with server
        FlutterVoipPushNotificationPlugin.didReceiveIncomingPush(with: payload, forType: type.rawValue)
    }
    
    // Handle invalidated push tokens
    func pushRegistry(_ registry: PKPushRegistry, didInvalidatePushTokenFor type: PKPushType) {
        FlutterVoipPushNotificationPlugin.didInvalidatePushToken(forType: type.rawValue)
    }
    
    // Handle incoming pushes
    func pushRegistry(_ registry: PKPushRegistry, didUpdate pushCredentials: PKPushCredentials, for type: PKPushType) {
        // Process the received push
        FlutterVoipPushNotificationPlugin.didUpdate(pushCredentials, forType: type.rawValue);
    }
}


ios/Classes/FlutterVoipPushNotificationPlugin.m
#import "FlutterVoipPushNotificationPlugin.h"

NSString *const FlutterVoipRemoteNotificationsRegistered = @"voipRemoteNotificationsRegistered";
NSString *const FlutterVoipPushTokenInvalidated = @"voipPushTokenInvalidated";
NSString *const FlutterVoipLocalNotificationReceived = @"voipLocalNotificationReceived";
NSString *const FlutterVoipRemoteNotificationReceived = @"voipRemoteNotificationReceived";

BOOL RunningInAppExtension(void)
{
    return [[[[NSBundle mainBundle] bundlePath] pathExtension] isEqualToString:@"appex"];
}

@implementation FlutterVoipPushNotificationPlugin {
    FlutterMethodChannel* _channel;
    BOOL _resumingFromBackground;
    PKPushRegistry * _voipRegistry;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterVoipPushNotificationPlugin* instance = [[FlutterVoipPushNotificationPlugin alloc] initWithRegistrar:registrar messenger:[registrar messenger]];
    [registrar addApplicationDelegate:instance];
}

- (instancetype)initWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar
                      messenger:(NSObject<FlutterBinaryMessenger>*)messenger{
    
    self = [super init];
    
    if (self) {
        _channel = [FlutterMethodChannel
                    methodChannelWithName:@"com.peerwaya/flutter_voip_push_notification"
                    binaryMessenger:[registrar messenger]];
        [registrar addMethodCallDelegate:self channel:_channel];
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(handleRemoteNotificationsRegistered:)
                                                     name:FlutterVoipRemoteNotificationsRegistered
                                                   object:nil];
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(handleLocalNotificationReceived:)
                                                     name:FlutterVoipLocalNotificationReceived
                                                   object:nil];
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(handleRemoteNotificationReceived:)
                                                     name:FlutterVoipRemoteNotificationReceived
                                                   object:nil];
        [[NSNotificationCenter defaultCenter] addObserver:self
                                                 selector:@selector(handlePushTokenInvalidated:)
                                                     name:FlutterVoipPushTokenInvalidated
                                                   object:nil];
    }
    return self;
}

- (void)dealloc
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    NSString *method = call.method;
    if ([@"requestNotificationPermissions" isEqualToString:method]) {
        if (RunningInAppExtension()) {
            result(nil);
            return;
        }
        [self registerUserNotification:call.arguments result:result];
    }if ([@"checkPermissions" isEqualToString:method]) {
        if (RunningInAppExtension()) {
            result(@{@"alert": @NO, @"badge": @NO, @"sound": @NO});
            return;
        }
        result([self checkPermissions]);
    }if ([@"presentLocalNotification" isEqualToString:method]) {
        [self presentLocalNotification:call.arguments];
        result(nil);
    }if ([@"getToken" isEqualToString:method]) {
        result([self getToken]);
    }else {
        result(FlutterMethodNotImplemented);
    }
}

- (void)registerUserNotification:(NSDictionary *)permissions result:(FlutterResult)result
{
    UIUserNotificationType notificationTypes = 0;
    if ([permissions[@"sound"] boolValue]) {
        notificationTypes |= UIUserNotificationTypeSound;
    }
    if ([permissions[@"alert"] boolValue]) {
        notificationTypes |= UIUserNotificationTypeAlert;
    }
    if ([permissions[@"badge"] boolValue]) {
        notificationTypes |= UIUserNotificationTypeBadge;
    }
    UIUserNotificationSettings *settings =
    [UIUserNotificationSettings settingsForTypes:notificationTypes categories:nil];
    [[UIApplication sharedApplication] registerUserNotificationSettings:settings];
    result(nil);
}

- (NSDictionary *)checkPermissions
{
    NSUInteger types = [[UIApplication sharedApplication] currentUserNotificationSettings].types;
    return @{
             @"alert": @((types & UIUserNotificationTypeAlert) > 0),
             @"badge": @((types & UIUserNotificationTypeBadge) > 0),
             @"sound": @((types & UIUserNotificationTypeSound) > 0),
             };
}

- (void)voipRegistration
{
#ifdef DEBUG
    NSLog(@"[FlutterVoipPushNotificationPlugin] voipRegistration");
#endif
    dispatch_queue_t mainQueue = dispatch_get_main_queue();
    // Create a push registry object
    _voipRegistry = [[PKPushRegistry alloc] initWithQueue: mainQueue];
    // Set the registry's delegate to self
    _voipRegistry.delegate = (FlutterVoipPushNotificationPlugin *)[UIApplication sharedApplication].delegate;
    // Set the push type to VoIP
    _voipRegistry.desiredPushTypes = [NSSet setWithObject:PKPushTypeVoIP];
}

- (void)presentLocalNotification:(UILocalNotification *)notification
{
    [[UIApplication sharedApplication] presentLocalNotificationNow:notification];
}

#pragma mark - AppDelegate

- (BOOL)application:(UIApplication *)application
    didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
  [self voipRegistration];
  return YES;
}

- (void)applicationDidEnterBackground:(UIApplication *)application {
  _resumingFromBackground = YES;
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
  _resumingFromBackground = NO;
}


- (NSString*)getToken
{
    NSMutableString *hexString = [NSMutableString string];
    NSData* token = [_voipRegistry pushTokenForType:PKPushTypeVoIP];
    NSUInteger voipTokenLength = token.length;
    const unsigned char *bytes = token.bytes;
    for (NSUInteger i = 0; i < voipTokenLength; i++) {
        [hexString appendFormat:@"%02x", bytes[i]];
    }
    return hexString;
}

#pragma mark - PKPushRegistryDelegate methods

+ (void)didUpdatePushCredentials:(PKPushCredentials *)credentials forType:(NSString *)type
{
#ifdef DEBUG
    NSLog(@"[FlutterVoipPushNotificationPlugin] didUpdatePushCredentials credentials.token = %@, type = %@", credentials.token, type);
#endif
    NSMutableString *hexString = [NSMutableString string];
    NSUInteger voipTokenLength = credentials.token.length;
    const unsigned char *bytes = credentials.token.bytes;
    for (NSUInteger i = 0; i < voipTokenLength; i++) {
        [hexString appendFormat:@"%02x", bytes[i]];
    }
    [[NSNotificationCenter defaultCenter] postNotificationName:FlutterVoipRemoteNotificationsRegistered
                                                        object:self
                                                      userInfo:@{@"deviceToken" : [hexString copy]}];
   
}

+ (void)didInvalidatePushTokenForType:(NSString *)type
{
#ifdef DEBUG
    NSLog(@"[FlutterVoipPushNotificationPlugin] didInvalidatePushTokenFor type = %@", type);
#endif
    [[NSNotificationCenter defaultCenter] postNotificationName:FlutterVoipPushTokenInvalidated
                                                        object:self];
}

+ (void)didReceiveIncomingPushWithPayload:(PKPushPayload *)payload forType:(NSString *)type
{
#ifdef DEBUG
    NSLog(@"[FlutterVoipPushNotificationPlugin] didReceiveIncomingPushWithPayload payload.dictionaryPayload = %@, type = %@", payload.dictionaryPayload, type);
#endif
    [[NSNotificationCenter defaultCenter] postNotificationName:FlutterVoipRemoteNotificationReceived
                                                        object:self
                                                      userInfo:payload.dictionaryPayload];
}

- (void)handleRemoteNotificationsRegistered:(NSNotification *)notification
{
#ifdef DEBUG
    NSLog(@"[FlutterVoipPushNotificationPlugin] handleRemoteNotificationsRegistered notification.userInfo = %@", notification.userInfo);
#endif
    [_channel invokeMethod:@"onToken" arguments:notification.userInfo];
}

- (void)handleLocalNotificationReceived:(NSNotification *)notification
{
#ifdef DEBUG
    NSLog(@"[FlutterVoipPushNotificationPlugin] handleLocalNotificationReceived notification.userInfo = %@",notification.userInfo);
#endif
    if (_resumingFromBackground) {
        [_channel invokeMethod:@"onResume" arguments:@{@"local": @YES, @"notification": notification.userInfo}];
    } else {
        [_channel invokeMethod:@"onMessage" arguments:@{@"local": @YES, @"notification": notification.userInfo}];
    }
}

- (void)handleRemoteNotificationReceived:(NSNotification *)notification
{
#ifdef DEBUG
    NSLog(@"[FlutterVoipPushNotificationPlugin] handleRemoteNotificationReceived notification.userInfo = %@", notification.userInfo);
#endif
    if (_resumingFromBackground) {
        [_channel invokeMethod:@"onResume" arguments:@{@"local": @NO, @"notification": notification.userInfo}];
    } else {
        [_channel invokeMethod:@"onMessage" arguments:@{@"local": @NO, @"notification": notification.userInfo}];
    }
}

- (void)handlePushTokenInvalidated:(NSNotification *)notification
{
#ifdef DEBUG
    NSLog(@"[FlutterVoipPushNotificationPlugin] handlePushTokenInvalidated notification.userInfo = %@", notification.userInfo);
#endif
    [_channel invokeMethod:@"onTokenInvalidated" arguments:@{@"deviceToken": [self getToken]}];
}

@end


android/src/main/kotlin/com/example/flutter_voip_push_notification/FlutterVoipPushNotificationPlugin.kt
package com.example.flutter_voip_push_notification

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar

class FlutterVoipPushNotificationPlugin: MethodCallHandler {
  companion object {
    @JvmStatic
    fun registerWith(registrar: Registrar) {
      val channel = MethodChannel(registrar.messenger(), "flutter_voip_push_notification")
      channel.setMethodCallHandler(FlutterVoipPushNotificationPlugin())
    }
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    if (call.method == "getPlatformVersion") {
      result.success("Android ${android.os.Build.VERSION.RELEASE}")
    } else {
      result.notImplemented()
    }
  }
}


lib/flutter_voip_push_notification.dart
import 'dart:async';

import 'package:flutter/services.dart';
import 'package:flutter/widgets.dart';

/// Message handler for incoming notification
///
/// [isLocal] is true if this is a notification from a message scheduled locally
/// or false if its a remote voip push notification
/// [message] contains the notification payload see link below for how to parse this data
/// https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/CreatingtheNotificationPayload.html#//apple_ref/doc/uid/TP40008194-CH10-SW1
typedef Future<dynamic> MessageHandler(
    bool isLocal, Map<String, dynamic> notification);

/// Handler for invalid PushKit tokens
///
/// [invalidToken] is the token that has been invalidated and should be removed from your server
/// https://stackoverflow.com/questions/46977380/voip-push-under-what-circumstances-does-didinvalidatepushtokenfortype-get-calle#47015401
typedef Future<dynamic> InvalidTokenHandler(String invalidToken);

class NotificationSettings {
  const NotificationSettings({
    this.sound = true,
    this.alert = true,
    this.badge = true,
  });

  NotificationSettings._fromMap(Map<String, bool> settings)
      : sound = settings['sound'],
        alert = settings['alert'],
        badge = settings['badge'];

  final bool sound;
  final bool alert;
  final bool badge;

  @visibleForTesting
  Map<String, dynamic> toMap() {
    return <String, bool>{'sound': sound, 'alert': alert, 'badge': badge};
  }

  @override
  String toString() => 'PushNotificationSettings ${toMap()}';
}

class LocalNotification {
  const LocalNotification({
    this.alertBody,
    this.alertAction,
    this.soundName,
    this.category,
    this.userInfo,
  })  : assert(alertBody != null),
        assert(alertAction != null);

  /// The message displayed in the notification alert.
  final String alertBody;

  /// The [action] displayed beneath an actionable notification. Defaults to "view";
  final String alertAction;

  /// The sound played when the notification is fired (optional).
  final String soundName;

  /// The category of this notification, required for actionable notifications (optional).
  final String category;

  /// An optional object containing additional notification data.
  final Map<String, dynamic> userInfo;

  @visibleForTesting
  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      'alertBody': alertBody,
      'alertAction': alertAction,
      'soundName': soundName,
      'category': category,
      'userInfo': userInfo
    };
  }

  @override
  String toString() => 'LocalNotification ${toMap()}';
}

class FlutterVoipPushNotification {
  factory FlutterVoipPushNotification() => _instance;

  @visibleForTesting
  FlutterVoipPushNotification.private(MethodChannel channel)
      : _channel = channel;

  static final FlutterVoipPushNotification _instance =
      FlutterVoipPushNotification.private(
          const MethodChannel('com.peerwaya/flutter_voip_push_notification'));

  final MethodChannel _channel;
  String _token;
  MessageHandler _onMessage;
  MessageHandler _onResume;
  InvalidTokenHandler _onInvalidToken;

  final StreamController<String> _tokenStreamController =
      StreamController<String>.broadcast();

  /// Fires when a new device token is generated.
  Stream<String> get onTokenRefresh {
    return _tokenStreamController.stream;
  }

  /// Sets up [MessageHandler] for incoming messages.
  void configure({
    MessageHandler onMessage,
    MessageHandler onResume,
    InvalidTokenHandler onInvalidToken,
  }) {
    _onMessage = onMessage;
    _onResume = onResume;
    _onInvalidToken = onInvalidToken;
    _channel.setMethodCallHandler(_handleMethod);
    //_channel.invokeMethod<void>('configure');
  }

  Future<dynamic> _handleMethod(MethodCall call) async {
    final Map map = call.arguments.cast<String, dynamic>();
    switch (call.method) {
      case "onToken":
        _token = map["deviceToken"];
        _tokenStreamController.add(_token);
        return null;
      case "onTokenInvalidated":
        final String invalidToken = map["deviceToken"];
        return _onInvalidToken?.call(invalidToken);
      case "onMessage":
        final bool isLocal = map["local"];
        final Map<String, dynamic> notification =
            map["notification"].cast<String, dynamic>();
        return _onMessage?.call(isLocal, notification);
      case "onResume":
        final bool isLocal = map["local"];
        final Map<String, dynamic> notification =
            map["notification"].cast<String, dynamic>();
        return _onResume?.call(isLocal, notification);
      default:
        throw UnsupportedError("Unrecognized JSON message");
    }
  }

  /// Returns the locally cached push token
  Future<String> getToken() async {
    return await _channel.invokeMethod<String>('getToken');
  }

  /// Prompts the user for notification permissions the first time
  /// it is called.
  Future<void> requestNotificationPermissions([
    NotificationSettings iosSettings = const NotificationSettings(),
  ]) async {
    _channel.invokeMethod<void>(
      'requestNotificationPermissions',
      iosSettings.toMap(),
    );
  }

  /// Schedules the local [notification] for immediate presentation.
  Future<void> presentLocalNotification(LocalNotification notification) async {
    await _channel.invokeMethod<void>(
      'presentLocalNotification',
      notification.toMap(),
    );
  }
}
